components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,354 @@
     }
 
-    /// Lazily adjust the thread pool's size
-    pub fn adjust_with(&mut self, new_size: usize) {
-        if self.size == new_size {
-            return;
-        }
-        // TODO: after tokio supports adjusting thread pool size(https://github.com/tokio-rs/tokio/issues/3329),
-        //   adapt it.
-        let pool = (self.replace_pool_rule)(new_size, &self.thread_name)
-            .expect("failed to create tokio runtime for backup worker.");
-        self.pool = Some(DaemonRuntime::from_runtime(pool));
-        self.size = new_size;
-        (self.after_adjust)(new_size);
-    }
-
-    pub fn block_on<F>(&self, f: F) -> F::Output
+    pub fn block_on<Fut>(&self, fut: Fut)
     where
-        F: Future,
+        Fut: Future<Output = ()> + Send + 'static,
     {
-        self.pool
-            .as_ref()
-            .expect("ResizableRuntime: please call adjust_with() before block_on()")
-            .0
-            .as_ref()
-            .unwrap()
-            .block_on(f)
+        let handle = self.handle();
+        handle.block_on(fut);
     }
-}
 
-#[derive(Clone)]
-pub struct ResizableRuntimeHandle {
-    inner: Arc<RwLock<ResizableRuntime>>,
-}
-
-impl ResizableRuntimeHandle {
-    pub fn new(runtime: ResizableRuntime) -> Self {
-        ResizableRuntimeHandle {
-            inner: Arc::new(RwLock::new(runtime)),
+    pub fn handle(&self) -> DeamonRuntimeHandle {
+        DeamonRuntimeHandle {
+            inner: Arc::downgrade(&self.current_runtime),
         }
     }
 
-    pub fn spawn<Fut>(&self, fut: Fut)
-    where
-        Fut: Future<Output = ()> + Send + 'static,
-    {
-        let inner = self.inner.read().unwrap();
-        inner.spawn(fut);
-    }
+    // TODO: after tokio supports adjusting thread pool size(https://github.com/tokio-rs/tokio/issues/3329),
+    //   adapt it.
+    pub fn adjust_with(&mut self, new_size: usize) {
+        if self.size == new_size {
+            return;
+        }
 
-    pub fn adjust_with(&self, new_size: usize) {
-        let mut inner = self.inner.write().unwrap();
-        inner.adjust_with(new_size);
-    }
+        let new_version = self.version + 1;
+        let thread_name = format!("{}-v{}", self.thread_prefix, new_version);
+        let new_runtime = (self.replace_pool_rule)(new_size, &thread_name)
+            .unwrap_or_else(|_| panic!("failed to create tokio runtime {}", thread_name));
 
-    pub fn block_on<F>(&self, f: F) -> F::Output
-    where
-        F: Future,
-    {
-        let inner = self.inner.read().unwrap();
-        inner.block_on(f)
+        let old_runtime: DeamonRuntime;
+        {
+            let mut runtime_guard = self.current_runtime.lock().unwrap();
+            if self.size == new_size || self.version >= new_version {
+                warn!(
+                    "Resize task dropped";
+                    "thread_name" => &thread_name,
+                    "new_size" => new_size
+                );
+                return;
+            }
```

Comment:
It makes ZERO SENSE to double check for `size == new_size` here, remember `adjust_with` is taking an **`&mut self`**, meaning NO ONE IS ABLE TO MODIFY `self.size` while `adjust_with` is not finished. Put all code irrelevant to the `runtime` OUTSIDE this scope.

Reply:
Oh, you are right TAT

Reply:
done